### PR TITLE
[RLlib] Adds LR scheduling ability to BC and MARWIL

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
## Description

Before this PR, BC and MARWIL would not pick up LR schedules because we'd never send timesteps to the learner.
After this PR, we do send timesteps there to update LR after gradient updates.

### CODE REVIEW

File: rllib/core/learner/learner_group.py [hunk 1]
```diff
@@ -269,18 +269,29 @@ def update(
...
-            episodes_refs:
-            timesteps:
+            episodes_refs: A list of Ray ObjectRefs to the episodes. If there are more
+                than one Learner workers, the list of episode refs is split amongst these and
+                one list shard is sent to each Learner.
+            timesteps: A dictionary of timesteps to pass to the Learners. This is usually used
+                for learning rate scheduling.
+            training_data: A TrainingData object to use for the update. If not provided,
+                a new TrainingData object will be created from the batch, batches, batch_refs,
+                episodes, and episodes_refs.
```
Comment (ArturNiederfahrenhorst) [inline rllib/core/learner/learner_group.py:294]: Not sure why these were empty before. Fixing in this PR. Although not directly related.

File: rllib/algorithms/algorithm.py [hunk 2]
```diff
@@ -3890,11 +3890,12 @@ def _compile_iteration_results(self, *, train_results, eval_results):
...
-            # `ENV_RUNNER_RESULTS/NUM_ENV_STEPS_SAMPLED_LIFETIME`.
-            results[NUM_ENV_STEPS_SAMPLED_LIFETIME] = results.get(
-                ENV_RUNNER_RESULTS, {}
-            ).get(NUM_ENV_STEPS_SAMPLED_LIFETIME, 0)
+            if results.get(NUM_ENV_STEPS_SAMPLED_LIFETIME) is None:
+                # Backward compatibility `NUM_ENV_STEPS_SAMPLED_LIFETIME` is now:
+                # `ENV_RUNNER_RESULTS/NUM_ENV_STEPS_SAMPLED_LIFETIME`.
+                results[NUM_ENV_STEPS_SAMPLED_LIFETIME] = results.get(
+                    ENV_RUNNER_RESULTS, {}
+                ).get(NUM_ENV_STEPS_SAMPLED_LIFETIME, 0)
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/algorithm.py:3932]: No need to always set this from env runner results.
Specifically, with BC, there are no env runner results but we set NUM_ENV_STEPS_SAMPLED_LIFETIME in the training step function.

File: rllib/algorithms/bc/tests/test_bc.py [hunk 3]
```diff
@@ -86,6 +89,60 @@ def test_bc_compilation_and_learning_from_offline_file(self):
...
+        base_path = Path(__file__).parents[3]
+        data_path = "local://" / base_path / data_path
+
+        config = (
+            BCConfig()
+            .environment(env="CartPole-v1")
+            .api_stack(
+                enable_rl_module_and_learner=True,
+                enable_env_runner_and_connector_v2=True,
+            )
```
Comment (kamil-kaczmarek) [inline rllib/algorithms/bc/tests/test_bc.py]: we don't need this anymore, new stack is enabled by default.

File: rllib/algorithms/bc/tests/test_bc.py [hunk 4]
```diff
@@ -86,6 +89,60 @@ def test_bc_compilation_and_learning_from_offline_file(self):
...
+            .evaluation(
+                evaluation_interval=3,
+                evaluation_num_env_runners=1,
+                evaluation_duration=5,
+                evaluation_parallel_to_training=True,
+            )
+            # Note, the `input_` argument is the major argument for the
+            # new offline API.
+            .offline_data(
+                input_=[data_path.as_posix()],
```
Comment (kamil-kaczmarek) [inline rllib/algorithms/bc/tests/test_bc.py:109]: why it's `input_` with `_`?

File: rllib/algorithms/bc/tests/test_bc.py [hunk 5]
```diff
@@ -86,6 +89,60 @@ def test_bc_compilation_and_learning_from_offline_file(self):
...
+        algo = config.build()
+
+        done = False
+        while not done:
+            results = algo.train()
+            assert results[NUM_ENV_STEPS_SAMPLED_LIFETIME] > 0
+            lr = results[LEARNER_RESULTS][DEFAULT_POLICY_ID][
+                "default_optimizer_learning_rate"
+            ]
+            ts = results[NUM_ENV_STEPS_SAMPLED_LIFETIME]
```
Comment (kamil-kaczmarek) [inline rllib/algorithms/bc/tests/test_bc.py]: tiny detail: you can run this above assert (L133), then reuse ts symbol for assert and if.

File: rllib/algorithms/bc/tests/test_bc.py [hunk 6]
```diff
@@ -86,6 +89,60 @@ def test_bc_compilation_and_learning_from_offline_file(self):
...
+            if ts < 3000:
+                # The learning rate should be linearly interpolated.
+                expected_lr = 0.001 + (ts / 3000) * (0.01 - 0.001)
+                self.assertAlmostEqual(lr, expected_lr, places=6)
+            else:
+                self.assertEqual(lr, 0.01)
```
Comment (kamil-kaczmarek) [inline rllib/algorithms/bc/tests/test_bc.py:135]: nice!

File: rllib/algorithms/bc/tests/test_bc.py [hunk 7]
```diff
@@ -86,6 +89,60 @@ def test_bc_compilation_and_learning_from_offline_file(self):
...
+                done = True
+
```
Comment (kamil-kaczmarek) [inline rllib/algorithms/bc/tests/test_bc.py:137]: Shall we add: ` algo.stop()`

File: rllib/core/learner/learner_group.py (hunk 1 above)
Comment (kamil-kaczmarek) [reply rllib/core/learner/learner_group.py:294]: nice, thx!

File: rllib/core/learner/learner_group.py [hunk 8]
```diff
@@ -269,18 +269,29 @@ def update(
...
+                one list shard is sent to each Learner.
             episodes: A list of Episodes to process and perform the update
                 for. If there are more than one Learner workers, the list of episodes
                 is split amongst these and one list shard is sent to each Learner.
-            episodes_refs:
-            timesteps:
+            episodes_refs: A list of Ray ObjectRefs to the episodes. If there are more
+                than one Learner workers, the list of episode refs is split amongst these and
+                one list shard is sent to each Learner.
+            timesteps: A dictionary of timesteps to pass to the Learners. This is usually used
```
Comment (kamil-kaczmarek) [inline rllib/core/learner/learner_group.py]: Can these be used for other use cases?

File: rllib/algorithms/marwil/marwil.py [hunk 9]
```diff
@@ -485,6 +493,11 @@ class (multi-/single-learner setup) and evaluation on
                 training_data=training_data,
                 minibatch_size=self.config.train_batch_size_per_learner,
                 num_iters=self.config.dataset_num_iters_per_learner,
+                timesteps={
+                    NUM_ENV_STEPS_SAMPLED_LIFETIME: self.metrics.peek(
+                        NUM_ENV_STEPS_SAMPLED_LIFETIME
+                    )
```
Comment (kamil-kaczmarek) [inline rllib/algorithms/marwil/marwil.py:498]: I'd consider adding `default=0` to peek.
Why do we use `NUM_ENV_STEPS_SAMPLED_LIFETIME` instead of `NUM_ENV_STEPS_TRAINED_LIFETIME` ?

File: rllib/algorithms/algorithm.py (hunk 2 above)
Comment (kamil-kaczmarek) [reply rllib/algorithms/algorithm.py:3932]: I would add it as comment in the code, I was looking for this exact info!

**Review by kamil-kaczmarek — CHANGES_REQUESTED** [review]
@ArturNiederfahrenhorst great!

Just details in comments.

Can you add marwil test, similar in style to what you implemented in `test_bc.py`?

File: rllib/algorithms/bc/tests/test_bc.py (hunk 3 above)
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/bc/tests/test_bc.py]: Nice. I took this from the test above (didn't test without). I'm removing it from the other test in test_bc as well then. Thanks!

File: rllib/algorithms/bc/tests/test_bc.py (hunk 4 above)
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/bc/tests/test_bc.py:109]: input is a built-in function in python  https://stackoverflow.com/a/20670757

File: rllib/algorithms/bc/tests/test_bc.py (hunk 5 above)
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/bc/tests/test_bc.py]: Yep, good call thanks!

File: rllib/core/learner/learner_group.py (hunk 8 above)
Comment (ArturNiederfahrenhorst) [reply rllib/core/learner/learner_group.py]: Sure! Any schedule I guess.

File: rllib/algorithms/marwil/marwil.py (hunk 9 above)
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/marwil/marwil.py:498]: Considered it but I'm against it because it can make for nasty silent errors.
Your algorithm may perform bad because you always default to peeking zero here and messing up your learning rate schedule with this. This would be hard to debug. On the other hand, we do deterministically always log NUM_ENV_STEPS_SAMPLED_LIFETIME right above.

File: rllib/algorithms/bc/tests/test_bc.py (hunk 4 above)
Comment (kamil-kaczmarek) [reply rllib/algorithms/bc/tests/test_bc.py:109]: Likely separate discussion; Let's try to make user-facing API simple. Looking at this [code](https://github.com/ray-project/ray/blob/master/rllib/offline/offline_data.py#L36-L40) we can change it to `inputs` or `input_paths`?

**Review by kamil-kaczmarek — APPROVED** [review]

**Comment by kamil-kaczmarek** [comment]
@ArturNiederfahrenhorst approved! Left one comment for your (RLlib team[?]) consideration.

QQ:
1. Why do we use `NUM_ENV_STEPS_SAMPLED_LIFETIME` instead of `NUM_ENV_STEPS_TRAINED_LIFETIME` in this use case? Trained steps seems to be better here.
2. Can you add marwil test? Similar to what you implemented in test_bc.py?

**Comment by ArturNiederfahrenhorst** [comment]
> Why do we use NUM_ENV_STEPS_SAMPLED_LIFETIME instead of NUM_ENV_STEPS_TRAINED_LIFETIME in this use case? Trained steps seems to be better here.

While I 100% agree (all LR schedules should be based on steps trained), this is not a BC/MARWIL thing.
The LR scheduling is baked into the base learner class and as such we'd either have to either create a special solution for BC/MARWIL by overriding it in this PR (not a good idea) OR we'd change it for all algorithms which I don't think we should do in this PR either. I'd love to do that at some point though.

> Can you add marwil test? Similar to what you implemented in test_bc.py?

Done!
